server/api-util/lineItemHelpers.js

```diff
@@ -0,0 +1,192 @@
+const Decimal = require('decimal.js');
+const has = require('lodash/has');
+const { types } = require('sharetribe-flex-sdk');
+const { Money } = types;
+
+const { convertMoneyToNumber, unitDivisor, convertUnitToSubUnit } = require('./currency');
+const { nightsBetween, daysBetween } = require('./dates');
+const LINE_ITEM_NIGHT = 'line-item/night';
+const LINE_ITEM_DAY = 'line-item/day';
+
+/** Helper functions for constructing line items*/
+
+/**
+ * Calculates lineTotal for lineItem based on quantity.
+ * The total will be `unitPrice * quantity`.
+ *
+ * @param {Money} unitPrice
+ * @param {int} quantity
+ *
+ * @returns {Money} lineTotal
+ */
+exports.calculateTotalPriceFromQuantity = (unitPrice, unitCount) => {
+  const numericPrice = convertMoneyToNumber(unitPrice);
+  const numericTotalPrice = new Decimal(numericPrice).times(unitCount).toNumber();
+  return new Money(
+    convertUnitToSubUnit(numericTotalPrice, unitDivisor(unitPrice.currency)),
+    unitPrice.currency
+  );
+};
+
+/**
+ * Calculates lineTotal for lineItem based on percentage.
+ * The total will be `unitPrice * (percentage / 100)`.
+ *
+ * @param {Money} unitPrice
+ * @param {int} percentage
+ *
+ * @returns {Money} lineTotal
+ */
+exports.calculateTotalPriceFromPercentage = (unitPrice, percentage) => {
+  const numericPrice = convertMoneyToNumber(unitPrice);
+  const numericTotalPrice = new Decimal(numericPrice)
+    .times(percentage)
+    .dividedBy(100)
+    .toNumber();
+  return new Money(
+    convertUnitToSubUnit(numericTotalPrice, unitDivisor(unitPrice.currency)),
+    unitPrice.currency
+  );
+};
+
+/**
+ * Calculates lineTotal for lineItem based on seats and units.
+ * The total will be `unitPrice * units * seats`.
+ *
+ * @param {Money} unitPrice
+ * @param {int} unitCount
+ * @param {int} seats
+ *
+ * @returns {Money} lineTotal
+ */
+exports.calculateTotalPriceFromSeats = (unitPrice, unitCount, seats) => {
```

Comment:
Can any of the params (unitPrice, unitCount or seats) be negative?

Reply:
As far as I understand from specs here (https://github.com/sharetribe/sharetribe-custom/blob/master/src/sharetribe/tegel/action/privileged_set_line_items.clj#L176) unitPrice and unitCount can be also negative but seats can't (maybe @kpuputti or @sktoiva can confirm). So maybe we should check that at least seats is not a negative value.